### PR TITLE
[mlir][Transforms] Dialect conversion: add missing argument materialization.

### PR DESCRIPTION
**Pull request by bchetioui**
When replacing a block argument, previously to #117513, we would automatically insert a N->1 argument materialization. After #117513, this is no longer the case for 1->1 mappings.

As a result, no materialization is added until `ReplaceBlockArgRewrite` is committed, where `findOrBuildReplacementValue` inserts a source materialization. The switch from an argument materialization to a source materialization causes legalization to fail.

[Here is an example reproducer](https://github.com/openxla/xla/blob/eb9d08bae564680ff465d772ceb70f4d84542e8c/xla/mlir_hlo/tests/Dialect/mhlo/hlo-legalize-to-linalg.mlir#L3016-L3031).

### CODE REVIEW

File: mlir/lib/Transforms/Utils/DialectConversion.cpp [hunk 1]
```diff
@@ -1375,12 +1375,18 @@ Block *ConversionPatternRewriterImpl::applySignatureConversion(
     // used as a replacement.
     auto replArgs =
         newBlock->getArguments().slice(inputMap->inputNo, inputMap->size);
+    auto insertPoint = OpBuilder::InsertPoint(newBlock, newBlock->begin());
     if (replArgs.size() == 1) {
-      mapping.map(origArg, replArgs.front());
+      // We need an argument materialization to replace the block argument.
+      Value argMat = buildUnresolvedMaterialization(
```
Comment (matthias-springer) [inline mlir/lib/Transforms/Utils/DialectConversion.cpp:1381]: This should not be necessary. Argument materialization are a workaround around missing 1:N support, but this is a 1:1 replacement.

**Review by matthias-springer — COMMENTED** [review]
> The switch from an argument materialization to a source materialization causes legalization to fail.

What kind of failure are you seeing? Is there no matching source materialization callback? It's possible that you have to turn an `addArgumentMaterialization` into an `addSourceMaterialization`.

**Comment by bchetioui** [comment]
> What kind of failure are you seeing? Is there no matching source materialization callback? It's possible that you have to turn an `addArgumentMaterialization` into an `addSourceMaterialization`.

This is the error I'm seeing:

```
error: failed to legalize unresolved materialization from ('f32') to ('tensor<f32>') that remained live after conversion
```

with IR

```
// -----// IR Dump After HloLegalizeToLinalgPass Failed (hlo-legalize-to-linalg) //----- //
func.func @map_mixed(%arg0: tensor<?xf32>, %arg1: tensor<4xf32>) -> tensor<?xf32> {
  %0 = shape.shape_of %arg0 : tensor<?xf32> -> tensor<1xindex>
  %c0 = arith.constant 0 : index
  %extracted = tensor.extract %0[%c0] : tensor<1xindex>
  %1 = tensor.empty(%extracted) : tensor<?xf32>
  %2 = linalg.generic {indexing_maps = [affine_map<(d0) -> (d0)>, affine_map<(d0) -> (d0)>, affine_map<(d0) -> (d0)>], iterator_types = ["parallel"]} ins(%arg0, %arg1 : tensor<?xf32>, tensor<4xf32>) outs(%1 : tensor<?xf32>) {
  ^bb0(%in: f32, %in_0: f32, %out: f32):
    %3 = builtin.unrealized_conversion_cast %in_0 : f32 to tensor<f32>
    %4 = builtin.unrealized_conversion_cast %in : f32 to tensor<f32>
    %extracted_1 = tensor.extract %4[] : tensor<f32>
    %extracted_2 = tensor.extract %3[] : tensor<f32>
    %5 = arith.addf %extracted_1, %extracted_2 : f32
    %from_elements = tensor.from_elements %5 : tensor<f32>
    %extracted_3 = tensor.extract %from_elements[] : tensor<f32>
    linalg.yield %extracted_3 : f32
  } -> tensor<?xf32>
  return %2 : tensor<?xf32>
}
```

I believe that the materialization is inserted [here](https://github.com/llvm/llvm-project/blob/aa2fdc69d35ff1c4a6de8a8d8edcc4c15236bb15/mlir/lib/Transforms/Utils/DialectConversion.cpp#L1479-L1484). But admittedly, I'm not that familiar with these internals---let me try to look into `addArgumentMaterialization`.

**Comment by zero9178** [comment]
> > What kind of failure are you seeing? Is there no matching source materialization callback? It's possible that you have to turn an `addArgumentMaterialization` into an `addSourceMaterialization`.
> 
> This is the error I'm seeing:
> 
> ```
> error: failed to legalize unresolved materialization from ('f32') to ('tensor<f32>') that remained live after conversion
> ```
> 
> with IR
> 
> ```
> // -----// IR Dump After HloLegalizeToLinalgPass Failed (hlo-legalize-to-linalg) //----- //
> func.func @map_mixed(%arg0: tensor<?xf32>, %arg1: tensor<4xf32>) -> tensor<?xf32> {
>   %0 = shape.shape_of %arg0 : tensor<?xf32> -> tensor<1xindex>
>   %c0 = arith.constant 0 : index
>   %extracted = tensor.extract %0[%c0] : tensor<1xindex>
>   %1 = tensor.empty(%extracted) : tensor<?xf32>
>   %2 = linalg.generic {indexing_maps = [affine_map<(d0) -> (d0)>, affine_map<(d0) -> (d0)>, affine_map<(d0) -> (d0)>], iterator_types = ["parallel"]} ins(%arg0, %arg1 : tensor<?xf32>, tensor<4xf32>) outs(%1 : tensor<?xf32>) {
>   ^bb0(%in: f32, %in_0: f32, %out: f32):
>     %3 = builtin.unrealized_conversion_cast %in_0 : f32 to tensor<f32>
>     %4 = builtin.unrealized_conversion_cast %in : f32 to tensor<f32>
>     %extracted_1 = tensor.extract %4[] : tensor<f32>
>     %extracted_2 = tensor.extract %3[] : tensor<f32>
>     %5 = arith.addf %extracted_1, %extracted_2 : f32
>     %from_elements = tensor.from_elements %5 : tensor<f32>
>     %extracted_3 = tensor.extract %from_elements[] : tensor<f32>
>     linalg.yield %extracted_3 : f32
>   } -> tensor<?xf32>
>   return %2 : tensor<?xf32>
> }
> ```
> 
> I believe that the materialization is inserted [here](https://github.com/llvm/llvm-project/blob/aa2fdc69d35ff1c4a6de8a8d8edcc4c15236bb15/mlir/lib/Transforms/Utils/DialectConversion.cpp#L1479-L1484). But admittedly, I'm not that familiar with these internals---let me try to look into `addArgumentMaterialization`.

Could you try adjusting this line: https://github.com/openxla/xla/blob/eb9d08bae564680ff465d772ceb70f4d84542e8c/xla/mlir_hlo/mhlo/utils/type_conversion.cc#L112
to also add the same function as a source materialization? 

As far as I can see in your IR, the only difference between before and now is that the `builtin.unrealized_conversion_cast` seen here is a source rather than argument materialization

**Comment by bchetioui** [comment]
> Could you try adjusting this line: https://github.com/openxla/xla/blob/eb9d08bae564680ff465d772ceb70f4d84542e8c/xla/mlir_hlo/mhlo/utils/type_conversion.cc#L112
to add the same function as a source materialization?

Thanks---I did manage to fix that test by changing the converter! (Though what fixed it was instead adding a target materialization---`addTargetMaterialization(scalarToTensor);`.)

At this point, I'm debugging other related failures where this trick doesn't seem to do (and for which I unfortunately don't have shareable reproducers). Nevertheless, I do expect that the issue is similar :)

Thank you @matthias-springer and @zero9178 for the assist, I'll close this now---since it seems unnecessary!

**Pull request closed without merging**